### DISCUSSION TITLE
Are subscriptions cached by default with graphcache ?

### DISCUSSION THREAD

**Discussion by efflam** [urql-graphql/urql#1997] [Q&A] [answered]
Hello,
I'm using `urql` with `next.js` thanks to `next-urql`.
`graphql-ws` is used for the subscription client.

I'm using `@urql/exchange-graphcache` like that: 

`const cache = cacheExchange({})`

And I'm calling `withUrqlClient` like that in my pages :

```
withUrqlClient(
		(ssrExchange) => ({
			url: process.env.NEXT_PUBLIC_GRAPHQL_ENDPOINT,
			exchanges: [
				devtoolsExchange,
				dedupExchange,
				cache,
				ssrExchange,
				authExchange(authExchangeConfig),
				fetchExchange,
				subscriptionExchange({
					forwardSubscription(operation) {
						return {
							subscribe: (sink) => {
								const dispose = wsClient?.subscribe(operation, sink);
								return {
									unsubscribe: dispose,
								};
							},
						};
					},
				}),
			],
		}),
		{ ssr: false },
	)(Component)
```

Using `urql dev tools`, I can see that Queries are cached as expected but data coming from the Subscriptions are not.

As a consequence,  I always see my loading state when I come back to the page with `useSubscription` as if I was using `requestPolicy: "network-only"` with `useQuery`.

From what I've read in the docs and past issues, exchanges should to be the correct order but I might be wrong.
Is there something more to do (like writing updaters in the cache configuration) ?

Thank you

**Comment by JoviDeCroock** [comment, marked as answer]
subscriptions and mutations are similar in the sense that their result is never cached it is used as an updater to the cache, this is why graphcache allows you to specify updaters for both of these

```js
cacheExchange({
  updates: {
    Subscription: {},
    Mutation: {},
  }
})
```

This means that for instance `mutation addTodo {}` and `subscription newTodo` could use the same updater (add a todo to a list). There is also no expectation to cache `subscriptions` as this is push hints coming from the server like notifications and hints about changed data as alluded to earlier. `fetching` in this case indicates that your subscription is acquiring a connection.

You can read more about this [here](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/)

**Reply by efflam** [reply]
OK thank for your answer.

> for mutations and subscriptions we cannot passively compute data

is clearly stated in the docs.

This post had let me thinking that cache updating was automatic :
https://spectrum.chat/urql/help/do-subscriptions-update-the-cache~cca02931-e8ca-4c1c-bc2f-e28e556664b0

**Reply by JoviDeCroock** [reply]
As also detailed in the docs, it _is_ automatic but _not always_ when you are listening for updates on a certain todo for instance, then we know that todo is already present in the cache but if you are listening for deleting/new things then we can't assume how this will impact the cache.

So generally the approach to take for custom-updates is, is it in cache? If it is then the update in the majority of cases will be automatic, if not then you will probably have to help the cache out. If it's a deletion as well.
